nexus/src/app/device_auth.rs

```diff
@@ -123,20 +123,67 @@ impl super::Nexus {
         let silo_max_ttl = silo_auth_settings.device_token_max_ttl_seconds;
         let requested_ttl = db_request.token_ttl_seconds;
 
-        // Validate the requested TTL against the silo's max TTL
-        if let (Some(requested), Some(max)) = (requested_ttl, silo_max_ttl) {
-            if requested > max.0.into() {
-                return Err(Error::invalid_request(&format!(
-                    "Requested TTL {} seconds exceeds maximum \
-                     allowed TTL for this silo of {} seconds",
-                    requested, max
-                )));
+        // This logic is a bit gnarly, but we landed on it as the least bad
+        // option. Error out if the user requests a token TTL that is longer
+        // than allowed, i.e., either
+        //
+        //   a) it is longer than the silo max TTL, or
+        //   b) this request was authenticated with a device token and the TTL
+        //      would produce an expiration later than the current token's.
+        //
+        // If the user does not request a specific TTL, we do not error out.
+        // We calculate the token TTL as min(silo max TTL, current token TTL
+        // if present). Token confirm requests authenticated with a console
+        // session can get device tokens with TTLs up to the silo max.
```

Comment:
We discussed this clamping logic in chat and decided that while it was more user-friendly to error out when TTL is specified, it is too easy to get the error when not specifying a TTL, so we clamp instead. Users confirming a token through a console session will never hit any of this logic anyway.